File: src/yourpkg/calculator.py
"""
Sequential calculator core logic.

This module implements the stateful calculator described in the design docs:
- sequential (standard) calculation model
- basic operators (+, -, *, /)
- repeated equals
- sign toggle
- clear / clear entry / backspace
- simple error handling (division by zero, invalid input)

Precision-specific logic (percent, tax, advanced rounding) will be layered on later tasks.
"""

from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal, InvalidOperation, getcontext
from typing import Optional


# Use a generous context; precision-specific rounding is handled separately.
getcontext().prec = 28

MAX_DIGITS = 12  # integer digit limit


@dataclass
class ErrorState:
    code: str
    message: str


class Calculator:
    def __init__(self) -> None:
        self.clear()

    def input_digit(self, digit: str) -> None:
        if self._is_error():
            return
        if digit not in "0123456789":
            self._set_error("INVALID_TOKEN", "Invalid digit")
            return

        if self.new_input:
            self.display_value = digit if digit != "0" else "0"
            self.new_input = False
        else:
            if self._digit_count(self.display_value) >= MAX_DIGITS:
                return
            if self.display_value == "0":
                self.display_value = digit
            else:
                self.display_value += digit

    def input_decimal_point(self) -> None:
        if self._is_error():
            return
        if self.new_input:
            self.display_value = "0."
            self.new_input = False
            return
        if "." not in self.display_value:
            self.display_value += "."

    def toggle_sign(self) -> None:
        if self._is_error():
            return
        if self.display_value.startswith("-"):
            self.display_value = self.display_value[1:]
        else:
            if self.display_value != "0":
                self.display_value = f"-{self.display_value}"

    def backspace(self) -> None:
        if self._is_error():
            return
        if self.new_input:
            return
        if len(self.display_value) <= 1 or (len(self.display_value) == 2 and self.display_value.startswith("-")):
            self.display_value = "0"
            self.new_input = True
        else:
            self.display_value = self.display_value[:-1]

    def clear_entry(self) -> None:
        if self._is_error():
            self.clear()
            return
        self.display_value = "0"
        self.new_input = True

    def clear(self) -> None:
        self.display_value: str = "0"
        self.accumulator: Optional[Decimal] = None
        self.pending_operator: Optional[str] = None
        self.recent_operand: Optional[Decimal] = None
        self.new_input: bool = True
        self.error: Optional[ErrorState] = None

    def set_operator(self, operator: str) -> None:
        if self._is_error():
            return
        if operator not in {"+", "-", "*", "/"}:
            self._set_error("INVALID_TOKEN", "Invalid operator")
            return

        current_value = self._to_decimal(self.display_value)
        if current_value is None:
            return

        if self.pending_operator is not None and self.accumulator is not None and not self.new_input:
            result = self._apply_operator(self.accumulator, current_value, self.pending_operator)
            if result is None:
                return
            self.accumulator = result
            self.display_value = self._format(result)
        elif self.accumulator is None:
            self.accumulator = current_value

        self.pending_operator = operator
        self.recent_operand = None
        self.new_input = True

    def equals(self) -> None:
        if self._is_error():
            return
        if self.pending_operator is None or self.accumulator is None:
            # nothing to do
            self.accumulator = self._to_decimal(self.display_value)
            self.new_input = True
            return

        operand = None
        if not self.new_input:
            operand = self._to_decimal(self.display_value)
            self.recent_operand = operand
        else:
            operand = self.recent_operand or self.accumulator

        if operand is None:
            return

        result = self._apply_operator(self.accumulator, operand, self.pending_operator)
        if result is None:
            return

        self.accumulator = result
        self.display_value = self._format(result)
        self.new_input = True

    def get_display(self) -> str:
        if self._is_error():
            return self.error.message
        return self.display_value

    # Internal helpers

    def _apply_operator(self, left: Decimal, right: Decimal, operator: str) -> Optional[Decimal]:
        try:
            if operator == "+":
                return left + right
            if operator == "-":
                return left - right
            if operator == "*":
                return left * right
            if operator == "/":
                if right == 0:
                    self._set_error("DIV_ZERO", "Cannot divide by zero")
                    return None
                return left / right
        except InvalidOperation:
            self._set_error("OVERFLOW", "Overflow")
            return None
        self._set_error("INVALID_TOKEN", "Invalid operator")
        return None

    def _to_decimal(self, value: str) -> Optional[Decimal]:
        try:
            return Decimal(value)
        except InvalidOperation:
            self._set_error("INVALID_TOKEN", "Invalid number")
            return None

    def _format(self, value: Decimal) -> str:
        # Normalize to plain string, strip trailing zeros.
        as_str = format(value, "f")
        if "." in as_str:
            as_str = as_str.rstrip("0").rstrip(".")
        return as_str if as_str else "0"

    def _digit_count(self, value: str) -> int:
        return sum(1 for ch in value if ch.isdigit())

    def _set_error(self, code: str, message: str) -> None:
        self.error = ErrorState(code=code, message=message)
        self.display_value = message
        self.pending_operator = None
        self.accumulator = None
        self.recent_operand = None
        self.new_input = True

    def _is_error(self) -> bool:
        return self.error is not None

File: src/lib/calculator.ts
export type Operator = "+" | "-" | "*" | "/";

export type ErrorState = {
  code: "DIV_ZERO" | "OVERFLOW" | "INVALID_TOKEN";
  message: string;
};

export type CalculatorState = {
  displayValue: string;
  accumulator: number | null;
  pendingOperator: Operator | null;
  recentOperand: number | null;
  newInput: boolean;
  error: ErrorState | null;
};

export const MAX_DIGITS = 12;

export function createInitialState(): CalculatorState {
  return {
    displayValue: "0",
    accumulator: null,
    pendingOperator: null,
    recentOperand: null,
    newInput: true,
    error: null,
  };
}

export function clearAll(state: CalculatorState): CalculatorState {
  return createInitialState();
}

export function clearEntry(state: CalculatorState): CalculatorState {
  if (state.error) {
    return createInitialState();
  }
  return {
    ...state,
    displayValue: "0",
    newInput: true,
  };
}

export function inputDigit(state: CalculatorState, digit: string): CalculatorState {
  if (state.error) return state;
  if (!/^[0-9]$/.test(digit)) {
    return setError(state, "INVALID_TOKEN", "Invalid digit");
  }

  if (state.newInput) {
    return { ...state, displayValue: digit === "0" ? "0" : digit, newInput: false };
  }

  const digits = countDigits(state.displayValue);
  if (digits >= MAX_DIGITS) return state;

  const next = state.displayValue === "0" ? digit : state.displayValue + digit;
  return { ...state, displayValue: next };
}

export function inputDecimal(state: CalculatorState): CalculatorState {
  if (state.error) return state;
  if (state.newInput) {
    return { ...state, displayValue: "0.", newInput: false };
  }
  if (!state.displayValue.includes(".")) {
    return { ...state, displayValue: state.displayValue + "." };
  }
  return state;
}

export function toggleSign(state: CalculatorState): CalculatorState {
  if (state.error) return state;
  if (state.displayValue === "0") return state;
  if (state.displayValue.startsWith("-")) {
    return { ...state, displayValue: state.displayValue.slice(1) };
  }
  return { ...state, displayValue: `-${state.displayValue}` };
}

export function backspace(state: CalculatorState): CalculatorState {
  if (state.error) return state;
  if (state.newInput) return state;
  if (state.displayValue.length <= 1 || (state.displayValue.startsWith("-") && state.displayValue.length <= 2)) {
    return { ...state, displayValue: "0", newInput: true };
  }
  return { ...state, displayValue: state.displayValue.slice(0, -1) };
}

export function setOperator(state: CalculatorState, operator: Operator): CalculatorState {
  if (state.error) return state;
  if (!["+", "-", "*", "/"].includes(operator)) {
    return setError(state, "INVALID_TOKEN", "Invalid operator");
  }

  const current = toNumber(state.displayValue);
  if (current == null) return state;

  let nextAccumulator = state.accumulator;
  let nextDisplay = state.displayValue;

  if (state.pendingOperator && state.accumulator != null && !state.newInput) {
    const result = applyOperator(state.accumulator, current, state.pendingOperator);
    if (result == null) return setError(state, "DIV_ZERO", "Cannot divide by zero");
    nextAccumulator = result;
    nextDisplay = formatNumber(result);
  } else if (state.accumulator == null) {
    nextAccumulator = current;
  }

  return {
    ...state,
    accumulator: nextAccumulator,
    pendingOperator: operator,
    recentOperand: null,
    newInput: true,
    displayValue: nextDisplay,
  };
}

export function equals(state: CalculatorState): CalculatorState {
  if (state.error) return state;
  if (state.pendingOperator == null || state.accumulator == null) {
    return { ...state, accumulator: toNumber(state.displayValue), newInput: true };
  }

  let operand: number | null = null;
  if (!state.newInput) {
    operand = toNumber(state.displayValue);
  } else {
    operand = state.recentOperand ?? state.accumulator;
  }
  if (operand == null) return state;

  const result = applyOperator(state.accumulator, operand, state.pendingOperator);
  if (result == null) return setError(state, "DIV_ZERO", "Cannot divide by zero");

  return {
    ...state,
    accumulator: result,
    displayValue: formatNumber(result),
    newInput: true,
    recentOperand: operand,
  };
}

// Helpers

function applyOperator(left: number, right: number, op: Operator): number | null {
  switch (op) {
    case "+":
      return left + right;
    case "-":
      return left - right;
    case "*":
      return left * right;
    case "/":
      if (right === 0) return null;
      return left / right;
    default:
      return null;
  }
}

function toNumber(value: string): number | null {
  const n = Number(value);
  return Number.isFinite(n) ? n : null;
}

function formatNumber(value: number): string {
  const asStr = value.toString();
  return asStr.includes(".") ? asStr.replace(/0+$/, "").replace(/\.$/, "") || "0" : asStr;
}

function countDigits(value: string): number {
  return (value.match(/[0-9]/g) || []).length;
}

function setError(state: CalculatorState, code: ErrorState["code"], message: string): CalculatorState {
  return {
    ...state,
    error: { code, message },
    displayValue: message,
    pendingOperator: null,
    accumulator: null,
    recentOperand: null,
    newInput: true,
  };
}
